File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int n = s.length(), i = 0, maxFreq = 0, windowSize = 0;
        int[] freq = new int[26];
        for(int j = 0; j < n; ++j) {
            freq[s.charAt(j) - 'A']++;
            maxFreq = Math.max(maxFreq, freq[s.charAt(j) - 'A']);
            windowSize = j - i + 1;
            if(windowSize > maxFreq + k) {
                freq[s.charAt(i) - 'A']--;
                i++;
            }
        }
        return n - i;
    }
}

File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        n, i, maxFreq, windowSize = len(s), 0, 0, 0
        freq = collections.Counter()
        for j in range (n):
            freq[s[j]] += 1
            maxFreq = max(maxFreq, freq[s[j]])
            windowSize = j - i + 1
            if windowSize > maxFreq + k:
                freq[s[i]] -= 1
                i += 1
        return n - i